src/Microsoft.Performance.Toolkit.Engine/PluginSet.cs

```diff
@@ -0,0 +1,375 @@
+        ///     in the given directories.
+        /// </summary>
+        /// <param name="extensionDirectories">
+        ///     The directories to search for plugins.
+        /// </param>
+        /// <returns>
+        ///     A new instance of the <see cref="PluginSet"/> class containing all
+        ///     of the successfully discovered plugins.
+        /// </returns>
+        /// <exception cref="System.ArgumentNullException">
+        ///     <paramref name="extensionDirectories"/> is <c>null</c>.
+        /// </exception>
+        /// <exception cref="System.ArgumentException">
+        ///     <paramref name="extensionDirectories"/> is empty.
+        /// </exception>
+        /// <exception cref="InvalidExtensionDirectoryException">
+        ///     One or more directory paths in <paramref name="extensionDirectories"/>
+        ///     is invalid or does not exist.
+        /// </exception>
+        public static PluginSet Load(
+            IEnumerable<string> extensionDirectories)
+        {
+            return Load(extensionDirectories, null);
+        }
+        /// <summary>
+        ///     Creates a new <see cref="PluginSet"/>, loading all plugins found
+        ///     in the given directories, using the given loading function.
+        /// </summary>
+        /// <param name="extensionDirectories">
+        ///     The directories to search for plugins.
+        /// </param>
+        /// <param name="assemblyLoader">
+        ///     The loader to use to load plugin assemblies. This parameter may be
+        ///     <c>null</c>. If this parameter is <c>null</c>, then the default
+        ///     loader will be used.
+        ///     <remarks>
+        ///         The default loader provides no isolation.
+        ///     </remarks>
+        /// </param>
+        /// <returns>
+        ///     A new instance of the <see cref="PluginSet"/> class containing all
+        ///     of the successfully discovered plugins. The returned instance will
+        ///     also contain a collection of non-fatal errors that occurred when
+        ///     creating this data set (i.e. a plugin failed to load.)
+        /// </returns>
+        /// <exception cref="System.ArgumentNullException">
+        ///     <paramref name="extensionDirectories"/> is <c>null</c>.
+        /// </exception>
+        /// <exception cref="System.ArgumentException">
+        ///     <paramref name="extensionDirectories"/> is empty.
+        /// </exception>
+        /// <exception cref="InvalidExtensionDirectoryException">
+        ///     One or more directory paths in <paramref name="extensionDirectories"/>
+        ///     is invalid or does not exist.
+        /// </exception>
+        public static PluginSet Load(
+            IEnumerable<string> extensionDirectories,
+            IAssemblyLoader assemblyLoader)
+        {
+            Guard.NotNull(extensionDirectories, nameof(extensionDirectories));
```

Comment:
Is there a reason you didn't want to use the PluginsLoader? We can isolate components more if we just make this PluginSet a consumer that passes directories down to the loader.

Reply:
I just moved the code out of the engine as it was. I can look at doing that.

Reply:
The pluginsloader is in an assembly that uses .NET Core, and the Engine is .NET standard. The Engine is .NET standard to facilitate use from many different places / scenarios, and there are no plans to upgrade it in the immediate / near future.